### PR TITLE
#220/create a database table to persist the chat room messages

### LINKED ISSUE #220: Create a database table to persist the chat room messages
A part of the user story: _As a chat room member, I want to be able to send messages and receive messages from other members in real time_

### PR DESCRIPTION
**Pull request by gbudau**
close #220

### CODE REVIEW

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { User } from '../user/user.domain';
+import { ChatRoomMessageWithUserEntity } from './infrastructure/db/chat-room-message-with-user.entity';
+
+export class ChatRoomMessageWithUser {
+  id: string;
+  user: User;
+  chatRoomId: string;
```
Comment (Damrod) [inline transcendence-app/src/chat/chat-room-message-with-user.domain.ts:7]: the name of this field is inconsistent with chatroom-message.domain.ts . `chatRoomId` vs `chatId`.

File: transcendence-app/src/user/infrastructure/db/user.entity.ts [hunk 2]
```diff
@@ -10,16 +10,38 @@ export enum userKeys {
   AVATAR_Y = '"avatarY"',
 }
 
+type UserEntityData = {
```
Comment (Damrod) [inline transcendence-app/src/user/infrastructure/db/user.entity.ts:13]: what do you need this type for?

File: transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+export enum ChatRoomMessageKeys {
+  ID = '"id"',
+  USER_ID = '"userId"',
+  CHAT_ROOM_ID = '"chatRoomId"',
+  CONTENT = '"content"',
+  CREATED_AT = '"createdAt"',
+}
+
+export interface ChatRoomMessageEntityData {
```
Comment (Damrod) [inline transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts:9]: Why do you need this model of splitting XXXData from XXXEntity with the same fields, and then building Entity from Data?

**Comment by Damrod** [comment]
I think we may need a bit more code to have context of how does this work.

**Review by Damrod — APPROVED** [review]
But I also trust you, so LGTM! Thank you!

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts (hunk 1 above)
Comment (vdedios) [reply transcendence-app/src/chat/chat-room-message-with-user.domain.ts:7]: I guess the chatId will never exist (individual chats will have a userId-userId relation) but agree to update it to chatRoomId so it's clearer what it refers to

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { User } from '../user/user.domain';
+import { ChatRoomMessageWithUserEntity } from './infrastructure/db/chat-room-message-with-user.entity';
+
+export class ChatRoomMessageWithUser {
```
Comment (vdedios) [inline transcendence-app/src/chat/chat-room-message-with-user.domain.ts:4]: Don't see a real value of having this. The idea of the domain is to validate the minimal pieces of our problem

We already have two separate domains for User and Chat. We may do this joined models in upper layers

File: transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts (hunk 3 above)
Comment (vdedios) [reply transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts:9]: Same doubt!

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts (hunk 1 above)
Comment (gbudau) [reply transcendence-app/src/chat/chat-room-message-with-user.domain.ts]: Good catch. I've fixed it.
`chatRoomId` is a good option too. For this PR, I preferred to rename it to `chatId` to match the names we already have in the `ChatMember` table.

File: transcendence-app/src/user/infrastructure/db/user.entity.ts (hunk 2 above)
Comment (gbudau) [reply transcendence-app/src/user/infrastructure/db/user.entity.ts:13]: We use this type to instantiate the CreateMessageWithUserEntity https://github.com/42AGV/ft_transcendence/blob/c463830a0d1a7e3914c087cb826b77234afbf7eb/transcendence-app/src/chat/infrastructure/db/chat-room-message-with-user.entity.ts#L22

File: transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts (hunk 3 above)
Comment (gbudau) [reply transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts:9]: When doing `SELECT * ...`, the node-pg library returns an array of JSON-like objects https://node-postgres.com/api/result#resultrows-arrayany

If we want to instantiate an entity from an object of the pg.Result.rows[i], we need to use the constructor.

Currently, we don't use the entities in our application (we use them like objects and don't use the constructors). For example, let's take a look at the user entity https://github.com/42AGV/ft_transcendence/blob/6740ec026f3b9e6e153d32a8fa5d50a8b0181b34/transcendence-app/src/user/infrastructure/db/user.entity.ts#L13-L24

The constructor takes around nine parameters, and if we want to create an `UserEntity` instance, we would have to pass them in order.

This PR adds the `ChatRoomMessageWithUserEntity`,  and for this, we need to use the constructor https://github.com/42AGV/ft_transcendence/blob/c463830a0d1a7e3914c087cb826b77234afbf7eb/transcendence-app/src/chat/infrastructure/db/chat-room-message-with-user.entity.ts#L22

In order to type the constructor of an XXXEntity, we create the XXXData interface.

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts (hunk 4 above)
Comment (gbudau) [reply transcendence-app/src/chat/chat-room-message-with-user.domain.ts:4]: On the `/chatroom` page, we will display the messages from a chat room, and each message contains information about the user (username and avatar)

How do you suggest we do these joined models in the upper layers?

**Comment by gbudau** [comment]
Thank you for your feedback!

File: transcendence-app/src/chat/chat-room-message-with-user.domain.ts (hunk 4 above)
Comment (gbudau) [reply transcendence-app/src/chat/chat-room-message-with-user.domain.ts:4]: I merge this as is to work on the other tasks that require this table. If there is a better way, we can change it later.

File: transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts (hunk 3 above)
Comment (vdedios) [reply transcendence-app/src/chat/infrastructure/db/chat-room-message.entity.ts:9]: Ok, so this is more like a data type workaround :+1:

**Pull request merged**